File: LeetCode/4Sum/main.swift
class Solution {
	func fourSum(_ arr: [Int], _ key: Int) -> [[Int]] {
		var ans = [[Int]]()
		let nums = arr.sorted()
		for i in nums.indices {
			for j in i + 1 ..< nums.count {
				let target = key - nums[i] - nums[j]
				var (lo, hi) = (j + 1, nums.count - 1)
				while lo < hi {
					let curr = nums[lo] + nums[hi]
					if curr == target {
						let temp = [nums[i], nums[j], nums[lo], nums[hi]]
						ans.append(temp)
						lo += 1
						hi -= 1
					} else if curr > target {
						hi -= 1
					} else {
						lo += 1
					}
				}
			}
		}
		return ans.makeUnique()
	}
}

extension Sequence where Iterator.Element: Hashable {
    func makeUnique() -> [Iterator.Element] {
        var ans = Set<Iterator.Element>()
        return self.filter { ans.insert($0).inserted }
    }
}

File: LeetCode/4Sum/main.cpp
class Solution {
   public:
    vector<vector<int>> fourSum(vector<int> &nums, int key) {
        vector<vector<int>> ans;
        int n = nums.size();
        sort(begin(nums), end(nums));
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                int target = key - nums[i] - nums[j];
                int lo = j + 1, hi = n - 1;
                while (lo < hi) {
                    int curr = nums[lo] + nums[hi];
                    if (curr == target) {
                        vector<int> temp = {nums[i], nums[j], nums[lo], nums[hi]};
                        ans.push_back(temp);
                        ++lo;
                        --hi;
                    } else if (curr > target) {
                        --hi;
                    } else {
                        ++lo;
                    }
                }
            }
        }
        auto makeUnique = [&](const vector<vector<int>> &arr) {
            set<vector<int>> ans(cbegin(arr), cend(arr));
            return vector<vector<int>>(cbegin(ans), cend(ans));
        };
        return makeUnique(ans);
    }
};